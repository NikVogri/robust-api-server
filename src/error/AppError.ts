interface IAppError {
    message: string;
    statusCode: number;
    thrownError?: Error | unknown;
    payload?: Record<string, unknown>;
}

export class AppError extends Error {
    thrownError?: Error | unknown;
    statusCode: number;
    payload?: Record<string, unknown>;

    constructor(public data: IAppError) {
        super(data.message);

        this.thrownError = data.thrownError;
        this.statusCode = data.statusCode;
        this.payload = data.payload;
    }

    static toStringifiableObject(error: Error | unknown) {
        if (!(error instanceof Error)) return {};

        const err: Record<string, unknown> = {};

        Object.getOwnPropertyNames(error).forEach((prop) => {
            if ((error[prop as keyof Error] as unknown) instanceof Error) {
                err[prop] = AppError.toStringifiableObject(error);
            } else err[prop] = error[prop as keyof Error];
        });

        return err;
    }
}
