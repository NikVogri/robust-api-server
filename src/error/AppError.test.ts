import { describe, expect, it } from 'vitest';
import { AppError } from './AppError';

describe('Test AppError', () => {
    it('should should initialize payload as expected', async () => {
        const error = new AppError({
            message: 'error message',
            statusCode: 400,
            payload: {
                some: 'property',
            },
            thrownError: new Error('Some thrown error'),
        });

        expect(JSON.stringify(error)).toMatchSnapshot();
    });

    it('should create a stringifiable error object', async () => {
        const error = new Error('Some error');
        error.stack = 'some-stack';

        expect(AppError.toStringifiableObject(error)).toMatchSnapshot();
    });
});
