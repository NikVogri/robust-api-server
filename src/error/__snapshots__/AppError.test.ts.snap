// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test AppError > should create a stringifiable error object 1`] = `
{
  "message": "Some error",
  "stack": "some-stack",
}
`;

exports[`Test AppError > should should initialize payload as expected 1`] = `"{"data":{"message":"error message","statusCode":400,"payload":{"some":"property"},"thrownError":{}},"thrownError":{},"statusCode":400,"payload":{"some":"property"}}"`;
